<a href="https://colab.research.google.com/github/adnan-mujagic/steam-recommender-system-using-implicitly-inferred-ratings/blob/main/steam_dataset_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Importing dependencies
numpy, pandas, and time are libraries commonly used for Machine Learning and data processing

In [6]:
import numpy as np
import pandas as pd
import time

## Importing dataset
Importing the dataset of interest, in this case it is the `steam-200k.csv` file located inside my Google Drive

In [ ]:
dataset = pd.read_csv("/content/drive/MyDrive/ML/Steam/steam-200k.csv", header = 0, names = ["user_id", "game_name", "action", "hours_played", "_"])

# take a few samples from the dataset
dataset.sample(5)

user_id                      game_name    action  hours_played  _
184759  110248234  Call of Duty Modern Warfare 3      play          17.0  0
136479  183838531                       Portal 2  purchase           1.0  0
26482   105877396                     Dwarfs F2P      play           1.0  0
1541     11373749        Trials 2 Second Edition      play           7.7  0
32724    86912006        Post Apocalyptic Mayhem  purchase           1.0  0

I will drop the last column since it is redundant

In [ ]:
dataset = dataset.drop(columns = ["_"])

dataset.sample(5)

user_id                            game_name    action  hours_played
148011   11970504               Styx Master of Shadows  purchase           1.0
127916  197190654               Euro Truck Simulator 2  purchase           1.0
47055    20464587                            Robocraft      play           0.6
39033   156297441    Company of Heroes Opposing Fronts  purchase           1.0
126185  186126772  Call of Duty Black Ops II - Zombies  purchase           1.0

## Getting more understanding of the data


In [ ]:
dataset.shape

(199999, 4)

In [ ]:
print("Number of users in dataset is:", dataset.user_id.unique().shape[0])
print("Number of games in dataset is:", dataset.game_name.unique().shape[0])

Number of users in dataset is: 12393
Number of games in dataset is: 5155


In [ ]:
dataset.action.value_counts()

purchase    129510
play         70489
Name: action, dtype: int64

For the generation of implicit reviews, I will be comparing a particular user's playing time to the average playing time of other players for the game in question. The score will be high if it exceedes the average, and low if it is below average.

Before I get into that, I will create some helper functions.

I noticed the dataset is not necessarily very clean. There are two actions, namely `play` and `purchase`, which I want to avoid. I only want the total playing time of a particular user to be recorded, even if they just purchased and never played.

In [ ]:
def clean_hours_played(input_dataframe):  
  clean_dataframe = pd.DataFrame(columns = ["user_id", "game_name", "hours_played"])
  
  unique_users = input_dataframe.user_id.unique()
  unique_users_count = unique_users.shape[0]
  processed_users = 0
  
  for user_id in unique_users:
    user_actions = input_dataframe.loc[input_dataframe.user_id == user_id]
    for game_name in user_actions.game_name.unique():
      user_actions_per_game = user_actions.loc[user_actions.game_name == game_name]
      if user_actions_per_game.shape[0] == 2:
        # if played, append actual playtime
        try:
          hours_played = user_actions_per_game.loc[user_actions_per_game.action == "play"].iloc[0].hours_played
        except:
          hours_played = 0
      else:
        # otherwise append 0 hours played as actual playtime
        hours_played = 0
      clean_dataframe = clean_dataframe.append({'user_id': user_id, 'game_name': game_name, 'hours_played': hours_played}, ignore_index = True)
    processed_users = processed_users + 1
    print("Processed ", processed_users, "/", unique_users_count, ".")

  return clean_dataframe

In [12]:
# clean_dataset = clean_hours_played(dataset) -> run this the first time to generate clean dataset
clean_dataset_export_path = "/content/drive/MyDrive/ML/Steam/steam-200k-clean.csv"
clean_dataset = pd.read_csv(clean_dataset_export_path)

clean_dataset.head()

user_id                   game_name  hours_played
0  151603712  The Elder Scrolls V Skyrim           0.0
1  151603712                   Fallout 4          87.0
2  151603712                       Spore          14.9
3  151603712           Fallout New Vegas          12.1
4  151603712               Left 4 Dead 2           8.9

In [13]:
clean_dataset.sample(10)

user_id                                    game_name  hours_played
75720   131348564                                      Nidhogg           1.0
69093   159966899                                Left 4 Dead 2           0.0
30483    56002861                Counter-Strike Condition Zero           0.0
47649    33865373                                        Osmos           0.3
12331    92393218              Command and Conquer Red Alert 3          37.0
17241    80779496                              Devil May Cry 4           7.4
117305  168324795  Call of Duty Modern Warfare 2 - Multiplayer           2.1
52388     8585433                                     Waveform           0.0
7504      1364546                                     Ricochet           0.0
10721     4325465                                     Ricochet           0.0

In [14]:
clean_dataset.shape

(128804, 3)

In [15]:
clean_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128804 entries, 0 to 128803
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   user_id       128804 non-null  int64  
 1   game_name     128804 non-null  object 
 2   hours_played  128804 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 2.9+ MB


In [11]:
clean_dataset.to_csv(clean_dataset_export_path, index = False)

## Implicit rating deduction formula
Now that I have a cleaner dataset, I can start developing the algorithm for implicit rating deduction.

### Version 1
Rating for user `u` and game `g` is:

```python
rating = 5 * Sigmoid((hours_played_u_g - average_hours_played_g) / average_hours_played_g)
```

This should generate ratings in the range `(0, 5)`.

This function is further explained in the following cells.

In [16]:
# Wrapper function to measure execution time of other functions
def time_benchmark(func):
  
    def wrap(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()
        elapsed = end - start
        print(func.__name__, "returned a result after:", elapsed, "seconds")
        return result
    return wrap

In [17]:
@time_benchmark
def get_average_hours_played_by_game(game_name):
  return clean_dataset.loc[clean_dataset.game_name == game_name].hours_played.mean()

In [18]:
get_average_hours_played_by_game("Far Cry 3")

get_average_hours_played_by_game returned a result after: 0.017309904098510742 seconds


24.52173913043478

In [19]:
get_average_hours_played_by_game("Counter-Strike Global Offensive")

get_average_hours_played_by_game returned a result after: 0.017796754837036133 seconds


228.59178470254955

In [20]:
def get_hours_played_by_user_and_game(user_id, game_name):
  try:
    return clean_dataset.loc[(clean_dataset.user_id == user_id) & (clean_dataset.game_name == game_name)].iloc[0].hours_played
  except:
    return 0

In [22]:
get_hours_played_by_user_and_game(126340495, "Saints Row IV")

48.0

In [23]:
get_hours_played_by_user_and_game(254906420, "Team Fortress 2")

1.2

In [24]:
def sigmoid(x):
  return 1 / (1 + np.exp(-x))

In [25]:
sigmoid(0)

0.5

In [26]:
sigmoid(3)

0.9525741268224334

In [27]:
sigmoid(-5)

0.0066928509242848554

In [28]:
# turn this on if you want -> @time_benchmark
def get_user_rating_for_game(user_id, game_name):
  hours_played_by_user = get_hours_played_by_user_and_game(user_id, game_name)
  average_hours_played_for_game = get_average_hours_played_by_game(game_name)
  if average_hours_played_for_game == 0.0:
    joy_coefficient = 0
  else:
    joy_coefficient = (hours_played_by_user - average_hours_played_for_game) / average_hours_played_for_game
  return sigmoid(joy_coefficient) * 5

In [29]:
clean_dataset.sample(10)

user_id                              game_name  hours_played
31113    60736966          Counter-Strike Condition Zero           0.0
25177      298950                            The Swapper           0.2
37423    52811164                  Counter-Strike Source           0.2
17847    24469287             Omerta - City of Gangsters           1.1
49173    86469479  Agarest - Ultimate Equipment Pack DLC           0.0
23946    10869571               Half-Life Opposing Force           0.0
122958   53901705                 Far Cry 3 Blood Dragon           0.0
57415   156675178           Kingdoms of Amalur Reckoning           1.5
39918    80128229                 Styx Master of Shadows           0.9
109985  281535206                                  Trove          10.7

Let's analyze this formula

In [30]:
user_id = 98649241
game_name = "Warframe"

In [32]:
get_hours_played_by_user_and_game(user_id, game_name)

83.0

In [31]:
get_average_hours_played_by_game(game_name)

get_average_hours_played_by_game returned a result after: 0.0421900749206543 seconds


31.965289256198346

In [33]:
get_user_rating_for_game(user_id, game_name)

get_average_hours_played_by_game returned a result after: 0.010271310806274414 seconds


4.157689811271309

In the previous three cells, I gave an example of how the ratings are generated. 

From the sample dataset, you can see that user with id `98649241`, spent `83.0` hours playing a game called `Warframe`.

In the cell below, the average playtime for `Warframe` is calculated to be approximately `31.96`, which means this user played the game for approximately `1.6` whole average play times for that game. I named this number a `joy_coefficient`. 

Joy coefficient is high if the user plays the game more than average, it is low when the user plays the game less than average, and it is 0 if the user plays the game an average amount of time.


This leads to the conclusion that the user likes the game quite a lot, in fact, the implicily inferred rating by feeding the `joy_coefficient`to the `5 * sigmoid` function  is `4.16 / 5`, as you can see in the plot below:

![Sigmoid](https://drive.google.com/uc?id=12krBswoap1c35m9Pp8pZfqxb1gWg5xIV)

Finally, I will iterate through the current dataset, and create one in which each row will have the following columns: `user_id`, `game_name` and `rating`, where the rating will be generated using the above explained funciton.

In [ ]:
def get_dataset_with_implicit_ratings(input_dataframe, debug = False, progress_update_percentage = 1):
  dataframe_with_ratings = pd.DataFrame(columns = ["user_id", "game_name", "rating"])
  progress = 0.
  number_of_rows = input_dataframe.shape[0]
  for idx, row in input_dataframe.iterrows():
    if debug and ((idx + 1) / number_of_rows) * 100 > progress + progress_update_percentage:
      progress = progress + progress_update_percentage
      print("Processing progress:", progress, "%")
    user_id, game_name = row.user_id, row.game_name
    dataframe_with_ratings = dataframe_with_ratings.append({"user_id": user_id, "game_name": game_name, "rating": get_user_rating_for_game(user_id, game_name)}, ignore_index = True)
  return dataframe_with_ratings

In [ ]:
# clean_dataset = pd.read_csv("/content/drive/MyDrive/ML/Steam/steam-200k-clean.csv")

# clean_dataset = clean_dataset.iloc[:100]

dataset_with_ratings = get_dataset_with_implicit_ratings(clean_dataset, debug = True)

In [ ]:
dataset_with_ratings.sample(5)

user_id                      game_name    rating
21346  244290943                Audition Online  3.263658
86166   62486194               Empire Total War  3.593875
31634   24872208  Counter-Strike Condition Zero  1.344707
46333  302652704                       Unturned  1.402747
60977  100741663                 Rome Total War  1.714285

In [ ]:
dataset_with_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128804 entries, 0 to 128803
Data columns (total 3 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   user_id    128804 non-null  object 
 1   game_name  128804 non-null  object 
 2   rating     128804 non-null  float64
dtypes: float64(1), object(2)
memory usage: 2.9+ MB


In [ ]:
# Export clean dataset and dataset with ratings

dataset_with_ratings_export_path = "/content/drive/MyDrive/ML/Steam/steam-200k-with-ratings.csv"

dataset_with_ratings.to_csv(dataset_with_ratings_export_path, index = False)

Now that I have my inferred ratings, it's time to get started with the algorithm. I will do that in a separate notebook. You can follow see the implementation details of the recommender system at this [link](https://github.com/adnan-mujagic/steam-recommender-system-using-implicitly-inferred-ratings/blob/main/recommender_system.ipynb).